src/test/java/org/elasticsearch/cluster/ClusterInfoServiceTests.java

```diff
@@ -0,0 +1,259 @@
+        final InternalClusterInfoService infoService = (InternalClusterInfoService) internalTestCluster.getInstance(ClusterInfoService.class, internalTestCluster.getMasterName());
+        InfoListener listener = new InfoListener();
+        infoService.addListener(listener);
+        ClusterInfo info = listener.get();
+        assertNotNull("info should not be null", info);
+        Map<String, DiskUsage> usages = info.getNodeDiskUsages();
+        Map<String, Long> shardSizes = info.getShardSizes();
+        assertNotNull(usages);
+        assertNotNull(shardSizes);
+        assertThat("some usages are populated", usages.values().size(), Matchers.equalTo(2));
+        assertThat("some shard sizes are populated", shardSizes.values().size(), greaterThan(0));
+        for (DiskUsage usage : usages.values()) {
+            logger.info("--> usage: {}", usage);
+            assertThat("usage has be retrieved", usage.getFreeBytes(), greaterThan(0L));
+        }
+        for (Long size : shardSizes.values()) {
+            logger.info("--> shard size: {}", size);
+            assertThat("shard size is greater than 0", size, greaterThan(0L));
+        }
+    }
+
+    @Test
+    public void testClusterInfoServiceInformationClearOnError() throws InterruptedException, ExecutionException {
+        internalCluster().startNodesAsync(2,
+                // manually control publishing
+                ImmutableSettings.builder().put(InternalClusterInfoService.INTERNAL_CLUSTER_INFO_UPDATE_INTERVAL, "60m").build())
+                .get();
+        createIndex("test");
+        ensureGreen("test");
+        InternalTestCluster internalTestCluster = internalCluster();
+        InternalClusterInfoService infoService = (InternalClusterInfoService) internalTestCluster.getInstance(ClusterInfoService.class, internalTestCluster.getMasterName());
+        InfoListener listener = new InfoListener();
+        infoService.addListener(listener);
+
+        // get one healthy sample
+        infoService.updateOnce();
+        ClusterInfo info = listener.get();
+        assertNotNull("failed to collect info", info);
+        assertThat("some usages are populated", info.getNodeDiskUsages().size(), Matchers.equalTo(2));
+        assertThat("some shard sizes are populated", info.getShardSizes().size(), greaterThan(0));
+
+
+        MockTransportService mockTransportService = (MockTransportService) internalCluster().getInstance(TransportService.class, internalTestCluster.getMasterName());
+
+        final AtomicBoolean timeout = new AtomicBoolean(false);
+        final Set<String> blockedActions = ImmutableSet.of(NodesStatsAction.NAME, NodesStatsAction.NAME + "[n]", IndicesStatsAction.NAME, IndicesStatsAction.NAME + "[s]");
+        for (DiscoveryNode node : internalTestCluster.clusterService().state().getNodes()) {
+            mockTransportService.addDelegate(node, new MockTransportService.DelegateTransport(mockTransportService.original()) {
```

Comment:
Can you add a comment for what this is doing? (ie, registering a delegate that will drop all the nodes stats and indices stats requests silently)